### PR TITLE
`AnnData` Conversion Notebook

### PR DESCRIPTION
**Pull request by srivarra**
**If you haven't already, please read through our [contributing guidelines](https://ark-analysis.readthedocs.io/en/latest/_rtd/contributing.html) before opening your PR**

**What is the purpose of this PR?**

## Main Features
Adds a notebook for converting the cell table to per-FOV `AnnData` objects. By default, they get saved to `data/<my-cohort>/anndata` in the notebook. In addition, it briefly covers viewing components of the `AnnData` object, and creating data pipelines with a small application for constructing spatial neighbors with `squidpy`.

- Added the conversion class which creates a `AnnData` object with:
	- Renamed spatial coordinates (`centroid-0` $\to$ `centroid_y`, `centroid-1` $\to$ `centroid_x`), stored in `obsm["spatial"]`.
	- All markers get stored in `var_names` / `X`.
	- All other columns get stored in `obs`.
	- Dropped `settings.CELL_SIZE` in favor of the generic `"area"` as a column name in `obs`, as we can have tables for many kinds of segmentations / observations (cell, fiber, ez_seg, etc...)
- Added a function to load all AnnData Zarr stores in a directory to an `AnnCollection`.
- Added a custom `datapipe` to iterate over the `AnnCollection` FOV by FOV.

## Documentation
- Added a brief `AnnData` section for `data_types.md` with figures.
- Added a `AnnData` section for `development.md`.

## Testing
- Adjusted `test_utils.make_cell_table` to have the expected ordering for Cell Tables. Added `n_cells` and `n_markers` to create different sized cell tables.
- Added a utility function to generate `AnnData` tables with various values of `n_obs`, and floating point and categorical column types. (`test_utils.generate_anndata_table` and `test_utils.generate_anncollection`)

## How did you implement your changes

- Used `Dask` to work with converting large cell tables.
- Used [`TorchData`](https://pytorch.org/data/beta/index.html) to create the `AnnData` `IterDataPipe`. This allows us to implement more intricate data pipelines with shuffling, mapping, and filtering functionality, and construct data loaders with [`DataLoader2`](https://pytorch.org/data/beta/dataloader2.html)
- Adjusted tests which utilized `test_utils.make_cell_table`.

## Remaining Issues
- Need a plan for structuring how we will write modules and functions for interacting with `AnnData` objects. We can try some stuff, and see what works best.

### CODE REVIEW

File: src/ark/utils/data_utils.py [hunk 1]
```diff
@@ -792,3 +805,199 @@ def stitch_images_by_shape(data_dir, stitched_dir, img_sub_folder=None, channels
...
+
+    # Move centroids from obs to obsm["spatial"]
+    obsm_dd = obs_dd[[settings.CENTROID_0, settings.CENTROID_1]].rename(columns={settings.CENTROID_0: "centroid_x", settings.CENTROID_1: "centroid_y"})
+    obs_dd = obs_dd.drop(columns=[settings.CENTROID_0, settings.CENTROID_1])
+
+    # Create the AnnData object
+    adata: AnnData = AnnData(X=X_dd, obs=obs_dd, obsm={"spatial": obsm_dd})
+
+    # Convert any extra string labels to categorical if it's beneficial.
+    adata.strings_to_categoricals()
```
Comment (alex-l-kong) [inline src/ark/utils/data_utils.py:853]: Is this inplace?

File: src/ark/utils/data_utils.py [hunk 2]
```diff
@@ -792,3 +805,199 @@ def stitch_images_by_shape(data_dir, stitched_dir, img_sub_folder=None, channels
...
+
+    adata.write_zarr(pathlib.Path(save_dir, f"{fov_id}.zarr"), chunks=(1000, 1000))
+    return pathlib.Path(save_dir, f"{fov_id}.zarr").as_posix()
+
+
+class ConvertToAnnData:
+    """ A class which converts the Cell Table `.csv` file to a series of `AnnData` objects,
+    one object per FOV.
+    
+    The default parameters stored in `.obs` parameters include
```
Comment (alex-l-kong) [inline src/ark/utils/data_utils.py:863]: Is this a lead-in to the Args section, or was there supposed to be more?

File: src/ark/utils/data_utils.py [hunk 3]
```diff
@@ -792,3 +805,199 @@ def stitch_images_by_shape(data_dir, stitched_dir, img_sub_folder=None, channels
...
+        obs_names = [settings.CELL_LABEL, settings.CELL_SIZE, *ct_columns[obs_index_start:].to_list(), *extra_obs_parameters]
+
+        # Use "area" as the default area id instead of settings.CELL_SIZE to account for
+        # non-cellular observations (ez_seg, fiber, etc...)
+        if ("area" in obs_names) and (settings.CELL_SIZE in obs_names):
+            obs_names.remove(settings.CELL_SIZE)
+        elif ("area" not in obs_names) and (settings.CELL_SIZE in obs_names):
+            cell_table = cell_table.rename(columns={settings.CELL_SIZE: "area"})
+            obs_names.remove(settings.CELL_SIZE)
+            obs_names.append("area")
```
Comment (alex-l-kong) [inline src/ark/utils/data_utils.py:912]: This control feels a bit messy. Can we instead do something like:

```
if settings.CELL_SIZE in obs_names:
    obs_names.remove(settings.CELL_SIZE)

    if "area" not in obs_names:
        cell_table = cell_table.rename(columns={settings.CELL_SIZE: "area"})
        obs_names.append("area")
```

File: src/ark/utils/data_utils.py [hunk 4]
```diff
@@ -792,3 +805,199 @@ def stitch_images_by_shape(data_dir, stitched_dir, img_sub_folder=None, channels
         current_img = stitched_data.loc['stitched_image', :, :, chan].values
         image_utils.save_image(os.path.join(stitched_subdir, chan + '_stitched' + file_ext),
                                current_img)
+
+
+@delayed
+def _convert_ct_fov_to_adata(fov_dd: dd.DataFrame, var_names: list[str], obs_names: list[str], save_dir=None) -> str:
```
Comment (alex-l-kong) [inline src/ark/utils/data_utils.py:811]: Will there ever be a case where `save_dir` is `None`? If so, it would throw an error for the following lines:

```
adata.write_zarr(pathlib.Path(save_dir, f"{fov_id}.zarr"), chunks=(1000, 1000))
return pathlib.Path(save_dir, f"{fov_id}.zarr").as_posix()
```

**Review by alex-l-kong — CHANGES_REQUESTED** [review]
Beautiful stuff overall. I think once the lab gets familiar with `AnnData`/`dask`/`zarr` they'll find it very easy to use. Mostly structural comments.

File: src/ark/utils/data_utils.py [hunk 5]
```diff
@@ -792,3 +805,199 @@ def stitch_images_by_shape(data_dir, stitched_dir, img_sub_folder=None, channels
...
+
+    # Extract the X matrix
+    X_dd: dd.DataFrame = fov_dd[var_names]
+    
+    # Extract the obs dataframe and convert the cell label to integer
+    obs_dd: dd.DataFrame = fov_dd[obs_names].astype({settings.CELL_LABEL: int, settings.FOV_ID: str})
+    obs_dd["cell_meta_cluster"] = pd.Categorical(obs_dd["cell_meta_cluster"].astype(str))
+
+    # Move centroids from obs to obsm["spatial"]
+    obsm_dd = obs_dd[[settings.CENTROID_0, settings.CENTROID_1]].rename(columns={settings.CENTROID_0: "centroid_x", settings.CENTROID_1: "centroid_y"})
```
Comment (camisowers) [inline src/ark/utils/data_utils.py:846]: Technically region props places (0,0) in the upper left corner of the image, with centroid-0 being distance along the y-axis (downwards) and centroid-1 being a regular x-axis position.  It's a very annoying and minor issue, but wondering your thoughts on the best way to make this fact clear to users.

File: templates/anndata_conversion.ipynb [hunk 6]
```diff
@@ -0,0 +1,784 @@
+{
```
Comment (camisowers) [inline templates/anndata_conversion.ipynb:1]: <p>Maybe include the nice graphic from PR #1073 with the AnnData matrices structure.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/1079/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (camisowers) [inline templates/anndata_conversion.ipynb:1]: <p>Can we get rid of this output?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/1079/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (camisowers) [inline templates/anndata_conversion.ipynb:1]: <p>Feel like sections 1.1 and 1.2 can get grouped down under like an "Optional Tutorial Section". It should be clear to the user what is a necessary cell to run (the ones above) and what is optional for them to understand the new data structure.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/1079/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (alex-l-kong) [inline templates/anndata_conversion.ipynb:1]: <p>We can move this section to the <code>data_types.md</code> file in the documentation. Specifically, explaining the columns, what's in <code>obs</code> vs <code>obsm</code>, etc. should be something the user should get a good understanding of before running this notebook.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/1079/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (alex-l-kong) [inline templates/anndata_conversion.ipynb:1]: <p>This should go in the <code>development.md</code> section. You can see an example of how we document <code>xarray</code>. On a side note, feel free to open a PR to add onto the existing <code>xarray</code> documentation there.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/1079/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (alex-l-kong) [inline templates/anndata_conversion.ipynb:1]: <p>Also belongs in <code>development.md</code>.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/1079/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

File: src/ark/utils/data_utils.py (hunk 1 above)
Comment (srivarra) [reply src/ark/utils/data_utils.py:853]: Yeah.

File: src/ark/utils/data_utils.py (hunk 4 above)
Comment (srivarra) [reply src/ark/utils/data_utils.py:811]: `save_dir` can never be `None`. I'll remove that default value.

File: src/ark/utils/data_utils.py (hunk 2 above)
Comment (srivarra) [reply src/ark/utils/data_utils.py:863]: Looks like I missed that bit of docs.

File: templates/anndata_conversion.ipynb (hunk 6 above)
Comment (srivarra) [reply templates/anndata_conversion.ipynb:1]: That sounds like a good idea, I'll add an optional section and more the non conversion stuff over.

File: templates/anndata_conversion.ipynb [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+    "# `AnnData` Conversion\n",
+    "\n",
+    "The purpose of this notebook is to convert the cell table to a [`AnnData`](https://anndata.readthedocs.io/en/latest/index.html) Object.\n",
+    "\n",
+    "`AnnData` stands for Annotated Data, and is a data structure well suited for single cell data. It is a multi-faceted object composed of matrices and DataFrames which can be used to efficiently store and interact with our data.\n",
+    "\n",
+    "The following is a representation of the `AnnData` object schema:\n",
+    "\n",
+    "<p align=\"center\">\n",
+    "  <img width=\"50%\" src=\"../docs/_images/Anndata%20Schema.svg\" alt=\"AnnData Schema\"/>\n",
```
Comment (camisowers) [inline templates/anndata_conversion.ipynb:16]: Can you see the image on your end? Seems like it's not showing up https://github.com/angelolab/ark-analysis/blob/anndata-conversion-fov0/templates/anndata_conversion.ipynb

**Review by camisowers — COMMENTED** [review]
Notebook looks much cleaner!

File: templates/anndata_conversion.ipynb (hunk 7 above)
Comment (srivarra) [reply templates/anndata_conversion.ipynb:16]: Oh, forgot to change the link to the png image.

File: src/ark/utils/data_utils.py (hunk 5 above)
Comment (srivarra) [reply src/ark/utils/data_utils.py]: Just added to the centroid docs about the origin of NumPy Arrays in the `data_types.md` file. Let me know your thoughts!

Comment (camisowers) [reply src/ark/utils/data_utils.py]: The documentation explains it well!

**Review by camisowers — APPROVED** [review]

**Review by alex-l-kong — APPROVED** [review]
Looks good, sorry for the late review

**Comment by srivarra** [comment]
@ngreenwald

File: templates/anndata_conversion.ipynb [hunk 8]
```diff
@@ -0,0 +1,207 @@
+{
```
Comment (ngreenwald) [inline templates/anndata_conversion.ipynb:1]: <p>Does the embedded image show up correctly in the actual notebook? Not showing up here, could just be an issue with reviewnb</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/1079/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

File: docs/_rtd/data_types.md [hunk 9]
```diff
@@ -90,3 +91,70 @@ The CSV should contain the following columns
 * `fov`: name of the FOV the cell comes from
 * `label`: the name of the segmentation label
 * A set of expression columns defining the properties of each cell desired for clustering
+
+---
+
```
Comment (ngreenwald) [inline docs/_rtd/data_types.md:96]: This looks great! Love the asthetics

File: docs/_rtd/data_types.md [hunk 10]
```diff
@@ -90,3 +91,70 @@ The CSV should contain the following columns
...
+Each of these components have specific use cases and will be described below:
+
+#### 1. X, var, obs
+
+<p align="center">
+  <img width="50%" alt="image" src="https://github.com/angelolab/ark-analysis/assets/8909315/a5011077-d350-4aab-b8f8-609b11087bba">
+</p>
+
+- `X` is a matrix of shape `(n_obs, n_vars)` where `n_obs` is the number of observations and `n_vars` is the number of variables.
```
Comment (ngreenwald) [inline docs/_rtd/data_types.md:127]: Even though this data structure is quite flexible, to start with we're going to be using it primarily to store the cell table. I think having specific examples for each matrix, like you do the latter half of the matrices, will be helpful when people are browsing. We can always update this in the future once we have multiple types of AnnData objects, but for now filling it in assuming it will be a cell table I think will make it easier to understand.

File: src/ark/utils/data_utils.py [hunk 11]
```diff
@@ -792,3 +805,219 @@ def stitch_images_by_shape(data_dir, stitched_dir, img_sub_folder=None, channels
...
+                 extra_obs_parameters: list[str] = None) -> None:
+        
+        io_utils.validate_paths(paths=cell_table_path)
+        
+        
+        # Read in the cell table
+        cell_table: dd.DataFrame = dd.read_csv(cell_table_path)
+        ct_columns = cell_table.columns
+        
+        # Get te marker column indices
```
Comment (ngreenwald) [inline src/ark/utils/data_utils.py:898]:
```suggestion
        # Get the marker column indices
```

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looks awesome, just a couple small comments

File: templates/anndata_conversion.ipynb (hunk 8 above)
Comment (srivarra) [reply templates/anndata_conversion.ipynb:1]: yeah, I think it's cause the image is being displayed with html, instead of markdown's native method.

Here's what it looks like when I open the notebook in Jupyter Lab.
image

**Review by ngreenwald — APPROVED** [review]
Looks good
